go/pkg/cosmos/txhistory.go

```diff
@@ -126,92 +162,416 @@ func Messages(msgs []sdk.Msg) []Message {
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to parse totalSend: %s", res.Result.TotalCount)
+		return nil, errors.Wrapf(err, "error getting tendermint sent tx history")
 	}
 
-	resTxs := res.Result.Txs
+	return res, nil
+}
 
-	res, _, err = c.tendermintClient.InfoApi.TxSearch(c.ctx).Query(fmt.Sprintf("\"transfer.recipient='%s'\"", address)).Page(int32(page)).PerPage(int32(pageSize)).OrderBy("\"desc\"").Execute()
+func (c *txHistoryState) nextRecipTxHistoryPage() (*client.TxSearchResponse, error) {
+	c.cursor.TendermintRecvdPage++
+	res, err := c.tendermintHistoryRequest(histTxTypeRecipient)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to get receive transactions")
+		return nil, errors.Wrapf(err, "error getting tendermint recipient tx history")
 	}
 
-	totalReceive, err := strconv.ParseFloat(res.Result.TotalCount, 64)
+	return res, nil
+}
+
+func buildSentTxsQuery(address string) string {
+	queryTemplate := `"message.sender='%s'"`
+	return fmt.Sprintf(queryTemplate, address)
+}
+
+func buildRecvdTxsQuery(address string) string {
+	queryTemplate := `"transfer.recipient='%s'"`
+	return fmt.Sprintf(queryTemplate, address)
+}
+
+// given 2 sets of tx history, remove duplicates from the second set (recipient txs)
+func uniqueify(sender []client.TxSearchResponseResultTxs, recip []client.TxSearchResponseResultTxs) ([]client.TxSearchResponseResultTxs, []client.TxSearchResponseResultTxs) {
+	m := make(map[string]bool, len(sender)+len(recip))
+	filteredRecip := make([]client.TxSearchResponseResultTxs, 0, len(recip))
+	for i := 0; i < len(sender); i++ {
+		tx := sender[i]
+		m[*tx.Hash] = true
+	}
+	for i := 0; i < len(recip); i++ {
+		tx := recip[i]
+		if _, ok := m[*tx.Hash]; !ok {
+			m[*tx.Hash] = true
+			filteredRecip = append(filteredRecip, tx)
+		}
+	}
+	return sender, filteredRecip
+}
+
+func (c *Cursor) encode() *string {
+	if bytes, err := json.Marshal(c); err != nil {
+		logger.Errorf("error marshaling cursor from %#v: %s", c, err)
+		return nil
+	} else {
+		encoded := base64.StdEncoding.EncodeToString(bytes)
+		return &encoded
+	}
+}
+
+func (c *Cursor) decode(b64 string) error {
+	str, err := base64.StdEncoding.DecodeString(b64)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to parse totalReceive: %s", res.Result.TotalCount)
+		return errors.Wrapf(err, "error decoding base64 cursor %s", b64)
 	}
 
-	resTxs = append(resTxs, res.Result.Txs...)
+	if err := json.Unmarshal([]byte(str), c); err != nil {
+		return errors.Wrapf(err, "error marshaling cursor from %#v", c)
+	}
+	return nil
+}
 
-	// filter out duplicate transactions
-	seen := make(map[string]bool)
-	uniqueTxs := []client.TxSearchResponseResultTxs{}
-	for _, t := range resTxs {
-		if _, ok := seen[*t.Hash]; !ok {
-			uniqueTxs = append(uniqueTxs, t)
-			seen[*t.Hash] = true
+// read tendermint tx history
+func (s *txHistoryState) readTxHistory() (*TxHistory, error) {
+	// if we have cursor, restore state
+	if s.cursorSerialized != "" {
```

Comment:
I would pull this logic out into `GetTxHistory`, we have the serialized cursor at that point that we can either decode or instantiate a new one to interact with throughout the rest of the logic. Then we don't need to have `cursorSerialized` in the struct either since we never use it for anything afterwards